### PR TITLE
Deserialized plots from project files and from templates do not preserve colors

### PR DESCRIPTION
**Pull request by chamshop**
Fixes KBIbiopharma/particle_data_viewer/issues/726

### CODE REVIEW

File: pybleau/app/plotting/plot_style.py [hunk 1]
```diff
@@ -114,16 +115,13 @@ class BaseXYPlotStyle(HasStrictTraits):
...
 
     def __init__(self, **traits):
+        super(BaseXYPlotStyle, self).__init__(**traits)
+        self.renderer_style_manager.view_klass = self.view_klass
+
         # Support passing the renderer_styles directly since the
         # renderer_style_manager is there just for UI layout:
-        renderer_styles = None
         if "renderer_styles" in traits:
             renderer_styles = traits.pop("renderer_styles")
```
Comment (jonathanrocher) [inline pybleau/app/plotting/plot_style.py:124]: What's the point of working on the traits dict after super has been called?

File: pybleau/app/plotting/plot_style.py [hunk 2]
```diff
@@ -390,8 +388,8 @@ def _get_specific_view_elements(self):
 
     # Traits listener methods -------------------------------------------------
 
-    @on_trait_change("color_palette, renderer_styles[]")
-    def update_renderer_colors(self):
+    @observe("color_palette", post_init=True)
+    def update_renderer_colors(self, event):
```
Comment (jonathanrocher) [inline pybleau/app/plotting/plot_style.py:392]: What is the purpose of adding the `post_init=True`? Also, you don't seem to be using this `event` argument. Remove?

File: pybleau/app/plotting/tests/test_dataframe_plot_factories.py [hunk 3]
```diff
@@ -380,7 +380,8 @@ def test_multi_line_1_curve_plot(self):
         self.assert_renderers_have_distinct_colors(plot)
 
     def test_multi_line_2_curve_plot(self):
-        renderer_styles = [LineRendererStyle(), LineRendererStyle()]
+        renderer_styles = [LineRendererStyle(color="blue"),
+                           LineRendererStyle(color="red")]
```
Comment (jonathanrocher) [inline pybleau/app/plotting/tests/test_dataframe_plot_factories.py:384]: The point of this test was that you didn't have to specify different colors, they would be generated by the code. Did we lose that with this version?

File: pybleau/app/plotting/tests/test_dataframe_plot_factories.py [hunk 4]
```diff
@@ -396,8 +397,9 @@ def test_multi_line_2_curve_plot(self):
         self.assert_renderers_have_distinct_colors(plot)
 
     def test_multi_line_3_curve_plot(self):
-        renderer_styles = [LineRendererStyle(), LineRendererStyle(),
-                           LineRendererStyle()]
+        renderer_styles = [LineRendererStyle(color="blue"),
+                           LineRendererStyle(color="red"),
+                           LineRendererStyle(color="green")]
```
Comment (jonathanrocher) [inline pybleau/app/plotting/tests/test_dataframe_plot_factories.py:402]: Same comment: the point of this test was that you didn't have to specify different colors, they would be generated by the code. Did we lose that with this version?

**Review by jonathanrocher — CHANGES_REQUESTED** [review]
I have a question about an update you made to 2 existing tests. Also, don't we need a new IO test to make sure that if a plot's renderer colors were overwritten, serializing and deserializing doesn't lose these custom colors.

File: pybleau/app/plotting/tests/test_dataframe_plot_factories.py (hunk 3 above)
Comment (chamshop) [reply pybleau/app/plotting/tests/test_dataframe_plot_factories.py:384]: Yes, we did lose this when there is more than one renderer. The renderers should be created with a color (or have it set afterwards) if they are to be different. It is the only way we can control what the colors are. I am fine with removing the last assert (`assert_renderers_have_distinct_colors`), now, and just having the `assert_valid_plot` test.

> The point of this test was that you didn't have to specify different colors, they would be generated by the code. Did we lose that with this version?

File: pybleau/app/plotting/tests/test_dataframe_plot_factories.py (hunk 4 above)
Comment (chamshop) [reply pybleau/app/plotting/tests/test_dataframe_plot_factories.py:402]: Same comment as above.

> Same comment: the point of this test was that you didn't have to specify different colors, they would be generated by the code. Did we lose that with this version?

File: pybleau/app/plotting/plot_style.py (hunk 1 above)
Comment (chamshop) [reply pybleau/app/plotting/plot_style.py:124]: Yes! Thanks for catching this. I think this was a change that didn't get completely reverted back when I was trying out something with the `renderer_styles`. Will revert.

> What's the point of working on the traits dict after super has been called?

File: pybleau/app/plotting/tests/test_dataframe_plot_factories.py (hunk 3 above)
Comment (jonathanrocher) [reply pybleau/app/plotting/tests/test_dataframe_plot_factories.py:384]: Thanks for clarifying. I still agree it's the right thing to do. I agree that then, we should then remove the call to `assert_renderers_have_distinct_colors`.

**Comment by jonathanrocher** [comment]
One more thought: isn't there an opportunity to refactor all of these identical blocks:
```
            num_renderer = len(renderer_styles)
            if num_renderer > 1:
                colors = generate_chaco_colors(num_renderer)
                for curr, renderer in enumerate(renderer_styles):
                    renderer.color = colors[curr]
```
into a utility function used over and over?

File: pybleau/app/plotting/plot_style.py (hunk 2 above)
Comment (chamshop) [reply pybleau/app/plotting/plot_style.py:395]: I did some testing, and the `post_init` isn't necessary now.

As I understand it for `observe`, the handler has to have `event` in the arguments list [link](https://docs.enthought.com/traits/traits_user_manual/notification.html#observe-handler).

> What is the purpose of adding the post_init=True? Also, you don't seem to be using this event argument. Remove?

**Comment by chamshop** [comment]
Yeah, I noticed this, too. I decided not to for a good reason, but will revisit it and see if I can work it out.

> One more thought: is there an opportunity to refactor all of these identical blocks:
> 
> ```
>             num_renderer = len(renderer_styles)
>             if num_renderer > 1:
>                 colors = generate_chaco_colors(num_renderer)
>                 for curr, renderer in enumerate(renderer_styles):
>                     renderer.color = colors[curr]
> ```
> 
> into a utility function used over and over?

**Pull request merged**